tempesta_fw/http_parser.c

```diff
@@ -1812,41 +1812,145 @@ __req_parse_accept(TfwHttpReq *req, unsigned char *data, size_t len)
 
 	__FSM_START(parser->_i_st);
 
+	__FSM_STATE(Req_I_WSAccept) {
+		if (IS_WS(c))
+			__FSM_I_MOVE(Req_I_WSAccept);
+		/* Fall through. */
+	}
+
 	__FSM_STATE(Req_I_Accept) {
-		TRY_STR("text/html", Req_I_Accept, Req_I_AcceptHtml);
-		TRY_STR("*/*", Req_I_Accept, Req_I_AcceptHtml);
+		TRY_STR("text", Req_I_Accept, Req_I_AfterText);
+		TRY_STR_INIT();
+		if (c == '*')
+			__FSM_I_MOVE(Req_I_StarSlash);
+		if (IS_TOKEN(c))
+			__FSM_I_JMP(Req_I_Type);
+		return CSTR_NEQ;
+	}
+
+	__FSM_STATE(Req_I_AfterText) {
+		if (c == '/')
+			__FSM_I_MOVE(Req_I_AfterTextSlash);
+
+		__FSM_I_MOVE(Req_I_Type);
+	}
+
+	__FSM_STATE(Req_I_AfterTextSlash) {
+		TRY_STR("html", Req_I_AfterTextSlash, Req_I_AcceptHtml);
 		TRY_STR_INIT();
-		__FSM_I_JMP(Req_I_AcceptOther);
+		__FSM_I_JMP(Req_I_Subtype);
+	}
+
+	__FSM_STATE(Req_I_StarSlash) {
+		if (c == '/')
+			__FSM_I_MOVE(Req_I_StarSlashStar);
+		return CSTR_NEQ;
+	}
+
+	__FSM_STATE(Req_I_StarSlashStar) {
+		if (c == '*')
+			__FSM_I_MOVE(Req_I_AcceptHtml2);
+		return CSTR_NEQ;
 	}
 
 	__FSM_STATE(Req_I_AcceptHtml) {
 		if (IS_WS(c) || c == ',' || c == ';' || IS_CRLF(c)) {
 			__set_bit(TFW_HTTP_B_ACCEPT_HTML, req->flags);
 			__FSM_I_JMP(I_EoT);
 		}
-		__FSM_I_MOVE(Req_I_AcceptOther);
+		__FSM_I_JMP(Req_I_Subtype);
+	}
+
+	__FSM_STATE(Req_I_AcceptHtml2) {
+		if (IS_WS(c) || c == ',' || c == ';' || IS_CRLF(c)) {
+			__set_bit(TFW_HTTP_B_ACCEPT_HTML, req->flags);
+			__FSM_I_JMP(I_EoT);
+		}
+		return CSTR_NEQ;
+	}
+
+	__FSM_STATE(Req_I_Type) {
+		__FSM_I_MATCH_MOVE(token, Req_I_Type);
+		c = *(p + __fsm_sz);
+		if (c == '/')
+			__FSM_I_MOVE_n(Req_I_Slash, __fsm_sz + 1);
+		return CSTR_NEQ;
+	}
+
+	__FSM_STATE(Req_I_Slash) {
+		if (IS_TOKEN(c))
+			__FSM_I_JMP(Req_I_Subtype);
+		if (c == '*')
+			__FSM_I_MOVE(I_EoT);
+		return CSTR_NEQ;
+	}
+
+	__FSM_STATE(Req_I_Subtype) {
+		__FSM_I_MATCH_MOVE(token, Req_I_Subtype);
+		c = *(p + __fsm_sz);
+		if (IS_WS(c) || c == ',' || c == ';')
+			__FSM_I_MOVE_n(I_EoT, __fsm_sz);
+		if (IS_CRLF(c)) {
+			return __data_off(p + __fsm_sz);
+		}
```

Comment:
These 5 lines are just equal `__FSM_I_JMP(I_EoT)`. Moreover, the direct just is better because it doesn't allow sequences like `,;` (you allow them since you eat the separators here and in `I_EoT`).

BTW I asked to add test case for multiple consequent separators in https://github.com/tempesta-tech/tempesta/pull/1351/files#r326119365 , but the new version still has no the test.